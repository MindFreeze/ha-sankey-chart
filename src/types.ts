import {
  ActionConfig,
  BaseActionConfig,
  HapticType,
  LovelaceCard,
  LovelaceCardConfig,
  LovelaceCardEditor,
} from 'custom-card-helpers';
import { HassEntity, HassServiceTarget } from 'home-assistant-js-websocket';
import { UNIT_PREFIXES, CONVERSION_UNITS } from './const';

export interface SankeyChartConfig extends LovelaceCardConfig {
  type: string;
  autoconfig?: {
    print_yaml?: boolean;
  };
  title?: string;
  sections?: SectionConfig[];
  convert_units_to?: '' | CONVERSION_UNITS;
  co2_intensity_entity?: string;
  gas_co2_intensity?: number;
  monetary_unit?: string;
  electricity_price?: number;
  gas_price?: number;
  unit_prefix?: '' | keyof typeof UNIT_PREFIXES;
  round?: number;
  height?: number;
  wide?: boolean;
  layout?: 'auto' | 'vertical' | 'horizontal';
  show_icons?: boolean;
  show_names?: boolean;
  show_states?: boolean;
  show_units?: boolean;
  energy_date_selection?: boolean;
  min_box_size?: number;
  min_box_distance?: number;
  throttle?: number;
  min_state?: number;
  static_scale?: number;
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
}

declare global {
  interface HTMLElementTagNameMap {
    'sankey-chart-editor': LovelaceCardEditor;
    'hui-error-card': LovelaceCard;
  }
}

export type BoxType = 'entity' | 'passthrough' | 'remaining_parent_state' | 'remaining_child_state';

export interface EntityConfig {
  entity_id: string;
  add_entities?: string[];
  subtract_entities?: string[];
  // @deprecated #100
  substract_entities?: string[];
  attribute?: string;
  type?: BoxType;
  children?: ChildConfigOrStr[];
  unit_of_measurement?: string; // for attribute
  color?: string;
  name?: string;
  icon?: string;
  color_on_state?: boolean;
  color_above?: string;
  color_below?: string;
  color_limit?: number;
  tap_action?: ActionConfigExtended;
  double_tap_action?: ActionConfigExtended;
  hold_action?: ActionConfigExtended;
  // @deprecated
  remaining?:
    | string
    | {
        name: string;
        color?: string;
      };
}

export type EntityConfigInternal = EntityConfig & {
  children: ChildConfigOrStr[];
  accountedState?: number;
  foundChildren?: string[];
};

export type EntityConfigOrStr = string | EntityConfig;

export type ChildConfig = {
  entity_id: string;
  connection_entity_id: string;
};

export type ChildConfigOrStr = string | ChildConfig;

export type ActionConfigExtended = ActionConfig | CallServiceActionConfig | MoreInfoActionConfig | ZoomActionConfig;

export interface MoreInfoActionConfig extends BaseActionConfig {
  action: 'more-info';
  entity?: string;
  data?: {
    entity_id?: string | [string];
  };
}

export interface ZoomActionConfig extends BaseActionConfig {
  action: 'zoom';
}

export interface CallServiceActionConfig extends BaseActionConfig {
  action: 'call-service';
  service: string;
  data?: {
    entity_id?: string | [string];
    [key: string]: unknown;
  };
  target?: HassServiceTarget;
  repeat?: number;
  haptic?: HapticType;
}

export interface SectionConfig {
  entities: EntityConfigOrStr[];
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
  sort_group_by_parent?: boolean;
  min_width?: number;
}

export interface Section {
  entities: EntityConfigInternal[];
  sort_by?: 'none' | 'state';
  sort_dir?: 'asc' | 'desc';
  sort_group_by_parent?: boolean;
  min_width?: number;
}

export interface Config extends SankeyChartConfig {
  layout: 'auto' | 'vertical' | 'horizontal';
  unit_prefix: '' | keyof typeof UNIT_PREFIXES;
  round: number;
  height: number;
  min_box_size: number;
  min_box_distance: number;
  min_state: number;
  sections: Section[];
}

export interface Connection {
  startY: number;
  startSize: number;
  endY: number;
  endSize: number;
  state: number;
  startColor?: string;
  endColor?: string;
  highlighted?: boolean;
}

export interface Box {
  config: EntityConfigInternal;
  entity: Omit<HassEntity, 'state'> & {
    state: string | number;
  };
  entity_id: string;
  state: number;
  unit_of_measurement?: string;
  children: ChildConfigOrStr[];
  color: string;
  size: number;
  top: number;
  extraSpacers?: number;
  connections: {
    parents: Connection[];
  };
  connectedParentState: number;
}

export interface SectionState {
  boxes: Box[];
  total: number;
  spacerSize: number;
  statePerPixel: number;
  config: Section;
  size: number;
}

export interface ConnectionState {
  parent: EntityConfigInternal;
  child: EntityConfigInternal;
  state: number;
  prevParentState: number;
  prevChildState: number;
  ready: boolean;
  calculating?: boolean;
  highlighted?: boolean;
  passthroughs: EntityConfigInternal[];
}

export interface NormalizedState {
  state: number;
  unit_of_measurement?: string;
}
