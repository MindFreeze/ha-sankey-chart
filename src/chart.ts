import { LitElement, html, TemplateResult, PropertyValues, CSSResultGroup } from 'lit';
import { styleMap } from 'lit/directives/style-map';
import { classMap } from 'lit/directives/class-map';
import { until } from 'lit/directives/until.js';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { customElement, property, state } from 'lit/decorators';
import { HomeAssistant } from 'custom-card-helpers'; // This is a community maintained npm module with common helper functions/types. https://github.com/custom-cards/custom-card-helpers

import type { Config, SectionState, Box, ConnectionState, EntityConfigInternal, NormalizedState } from './types';
import { localize } from './localize/localize';
import styles from './styles';
import { getEntityId, normalizeStateValue, renderError, sortBoxes } from './utils';
import { HassEntities, HassEntity } from 'home-assistant-js-websocket';
import { handleAction } from './handle-actions';
import { filterConfigByZoomEntity } from './zoom';
import { renderSection } from './section';

@customElement('sankey-chart-base')
export class Chart extends LitElement {
  // https://lit.dev/docs/components/properties/
  @property({ attribute: false }) public hass!: HomeAssistant;
  @property({ attribute: false }) public states!: HassEntities;
  @property({ attribute: false }) public forceUpdateTs?: number;

  @state() private config!: Config;
  @state() private sections: SectionState[] = [];
  @state() private entityIds: string[] = [];
  @state() private connections: ConnectionState[] = [];
  @state() private connectionsByParent: Map<EntityConfigInternal, ConnectionState[]> = new Map();
  @state() private connectionsByChild: Map<EntityConfigInternal, ConnectionState[]> = new Map();
  @state() private statePerPixelY = 0;
  @state() private entityStates: Map<EntityConfigInternal | string, NormalizedState> = new Map();
  @state() private highlightedEntities: EntityConfigInternal[] = [];
  @state() private lastUpdate = 0;
  @state() public zoomEntity?: EntityConfigInternal;
  @state() public error?: Error;

  // https://lit.dev/docs/components/lifecycle/#reactive-update-cycle-performing
  protected shouldUpdate(changedProps: PropertyValues): boolean {
    if (!this.config) {
      return false;
    }
    if (
      changedProps.has('config') ||
      changedProps.has('forceUpdateTs') ||
      changedProps.has('highlightedEntities') ||
      changedProps.has('zoomEntity')
    ) {
      return true;
    }
    const now = Date.now();
    if (this.config.throttle && now - this.lastUpdate < this.config.throttle) {
      // woah there
      const ts = this.lastUpdate;
      setTimeout(() => {
        if (ts === this.lastUpdate) {
          // trigger manual update if no changes since last rejected update
          this.requestUpdate();
        }
      }, now - this.lastUpdate);
      return false;
    }

    const oldStates = changedProps.get('states') as HomeAssistant | undefined;
    if (!oldStates) {
      return false;
    }
    if (!Object.keys(oldStates).length) {
      return true;
    }
    return this.entityIds.some(id => {
      return oldStates[id] !== this.states[id] && oldStates[id].state !== this.states[id].state;
    });
  }

  public willUpdate(changedProps: PropertyValues): void {
    if (!this.entityIds.length || changedProps.has('config')) {
      this.entityIds = [];
      this.connections = [];
      this.connectionsByParent.clear();
      this.connectionsByChild.clear();
      this.config.sections.forEach(({ entities }, sectionIndex) => {
        entities.forEach(ent => {
          if (ent.type === 'entity') {
            this.entityIds.push(ent.entity_id);
          }
          ent.children.forEach(childConf => {
            const child = this.config.sections[sectionIndex + 1]?.entities.find(
              e => e.entity_id === getEntityId(childConf),
            );
            if (!child) {
              this.error = new Error(localize('common.missing_child') + ' ' + getEntityId(childConf));
              throw this.error;
            }
            const connection: ConnectionState = {
              parent: ent,
              child,
              state: 0,
              prevParentState: 0,
              prevChildState: 0,
              ready: false,
            };
            this.connections.push(connection);
            if (!this.connectionsByParent.has(ent)) {
              this.connectionsByParent.set(ent, []);
            }
            this.connectionsByParent.get(ent)!.push(connection);
            if (!this.connectionsByChild.has(child)) {
              this.connectionsByChild.set(child, []);
            }
            this.connectionsByChild.get(child)!.push(connection);
          });
        });
      });
    }
  }

  private _calcConnections() {
    const accountedIn = new Map<EntityConfigInternal, number>();
    const accountedOut = new Map<EntityConfigInternal, number>();
    this.connections.forEach(c => {
      c.ready = false;
    });
    this.connections.forEach(c => this._calcConnection(c, accountedIn, accountedOut));
  }

  private _calcConnection(
    connection: ConnectionState,
    accountedIn: Map<EntityConfigInternal, number>,
    accountedOut: Map<EntityConfigInternal, number>,
    force?: boolean,
  ) {
    if (connection.ready && !force) {
      return;
    }
    const { parent, child } = connection;
    [parent, child].forEach(ent => {
      if (ent.type === 'remaining_child_state') {
        this.connectionsByParent.get(ent)!.forEach(c => {
          if (!c.ready) {
            this.connectionsByChild.get(c.child)?.forEach(conn => {
              if (conn.parent !== parent) {
                this._calcConnection(conn, accountedIn, accountedOut);
              }
            });
          }
        });
      }
      if (ent.type === 'remaining_parent_state') {
        this.connectionsByChild.get(ent)?.forEach(c => {
          if (!c.ready) {
            this.connectionsByParent.get(c.parent)?.forEach(conn => {
              if (conn.child !== child) {
                this._calcConnection(conn, accountedIn, accountedOut);
              }
            });
          }
        });
      }
    });

    const parentStateNormalized = this._getMemoizedState(parent);
    const parentStateFull = parentStateNormalized.state ?? 0;
    connection.prevParentState = accountedOut.get(parent) ?? 0;
    const parentState = Math.max(0, parentStateFull - connection.prevParentState);
    const childStateNormalized = this._getMemoizedState(child);
    const childStateFull = childStateNormalized.state ?? 0;
    connection.prevChildState = accountedIn.get(child) ?? 0;
    const childState = Math.max(0, childStateFull - connection.prevChildState);

    if (!parentState || !childState) {
      connection.state = 0;
    } else {
      const connConfig = parent.children.find(c => getEntityId(c) === child.entity_id);
      if (typeof connConfig === 'object' && connConfig.connection_entity_id) {
        const connectionState = this._getMemoizedState(connConfig.connection_entity_id).state ?? 0;
        connection.state = Math.min(parentState, childState, connectionState);
      } else {
        connection.state = Math.min(parentState, childState);
      }
      accountedOut.set(parent, connection.prevParentState + connection.state);
      accountedIn.set(child, connection.prevChildState + connection.state);
    }
    connection.ready = true;
    if (
      (child.type === 'remaining_parent_state' &&
        (child.add_entities?.length || child.subtract_entities?.length) &&
        childState === Infinity) ||
      (parent.type === 'remaining_child_state' &&
        (parent.add_entities?.length || parent.subtract_entities?.length) &&
        parentState === Infinity)
    ) {
      // #111 remaining state with add/subtract entities
      accountedOut.set(parent, connection.prevParentState);
      accountedIn.set(child, connection.prevChildState);
      this._calcConnection(connection, accountedIn, accountedOut, true);
    }
    if (child.type === 'passthrough') {
      this.entityStates.delete(child);
    }
  }

  private _getMemoizedState(entityConfOrStr: EntityConfigInternal | string) {
    if (!this.entityStates.has(entityConfOrStr)) {
      const entityConf =
        typeof entityConfOrStr === 'string' ? { entity_id: entityConfOrStr, children: [] } : entityConfOrStr;
      const entity = this._getEntityState(entityConf);
      const unit_of_measurement = this._getUnitOfMeasurement(entityConf.unit_of_measurement || entity.attributes.unit_of_measurement);
      const normalized = normalizeStateValue(this.config.unit_prefix, Number(entity.state), unit_of_measurement);

      if (entityConf.type === 'passthrough') {
        const connections = this.connectionsByChild.get(entityConf);
        if (!connections) {
          throw new Error('Invalid entity config ' + JSON.stringify(entityConf));
        }
        const state = connections.reduce((sum, c) => (c.ready ? sum + c.state : Infinity), 0);
        if (state !== Infinity) {
          normalized.state = state;
        }
      }
      if (entityConf.add_entities) {
        entityConf.add_entities.forEach(subId => {
          const subEntity = this._getEntityState({ entity_id: subId, children: [] });
          const { state } = normalizeStateValue(
            this.config.unit_prefix,
            Number(subEntity.state),
            subEntity.attributes.unit_of_measurement || unit_of_measurement,
          );
          normalized.state += state;
        });
      }
      if (entityConf.subtract_entities) {
        entityConf.subtract_entities.forEach(subId => {
          const subEntity = this._getEntityState({ entity_id: subId, children: [] });
          const { state } = normalizeStateValue(
            this.config.unit_prefix,
            Number(subEntity.state),
            subEntity.attributes.unit_of_measurement || unit_of_measurement,
          );
          // stay positive
          normalized.state -= Math.min(state, normalized.state);
        });
      }
      if (normalized.state === Infinity) {
        // don't cache infinity
        return normalized;
      }
      this.entityStates.set(entityConfOrStr, normalized);
    }
    return this.entityStates.get(entityConfOrStr)!;
  }

  private _calcBoxes() {
    this.statePerPixelY = 0;
    if (this.config.static_scale) {
      // use static scale to set a minimum statePerPixelY
      this._calcBoxHeights(
        [{ state: this.config.static_scale, size: 0 } as Box],
        this.config.height,
        this.config.static_scale,
      );
    }
    const filteredConfig = filterConfigByZoomEntity(this.config, this.zoomEntity);
    const sectionsStates: SectionState[] = [];
    filteredConfig.sections.forEach(section => {
      let total = 0;
      const boxes: Box[] = section.entities
        .filter(entityConf => {
          const { min_state } = this.config;
          // remove empty entity boxes
          const { state } = this._getMemoizedState(entityConf);
          return state && state >= min_state;
        })
        .map(entityConf => {
          const { state, unit_of_measurement } = this._getMemoizedState(entityConf);
          total += state;

          let finalColor = entityConf.color || 'var(--primary-color)';
          if (typeof entityConf.color_on_state != 'undefined' && entityConf.color_on_state) {
            const colorLimit = typeof entityConf.color_limit === 'undefined' ? 1 : entityConf.color_limit;
            const colorBelow =
              typeof entityConf.color_below === 'undefined' ? 'var(--primary-color)' : entityConf.color_below;
            const colorAbove =
              typeof entityConf.color_above === 'undefined' ? 'var(--paper-item-icon-color)' : entityConf.color_above;
            if ( state > colorLimit ) {
              finalColor = colorAbove;
            } else if ( state < colorLimit ) {
              finalColor = colorBelow;
            }
          }

          return {
            config: entityConf,
            entity: this._getEntityState(entityConf),
            entity_id: getEntityId(entityConf),
            state,
            unit_of_measurement,
            color: finalColor,
            children: entityConf.children,
            connections: { parents: [] },
            top: 0,
            size: 0,
          };
        });
      if (!boxes.length) {
        return;
      }
      // leave room for margin
      const availableHeight = this.config.height - (boxes.length - 1) * this.config.min_box_distance;
      // calc sizes to determine statePerPixelY ratio and find the best one
      const calcResults = this._calcBoxHeights(boxes, availableHeight, total);
      const parentBoxes = section.sort_group_by_parent ? sectionsStates[sectionsStates.length - 1]?.boxes || [] : [];
      const sortBy = section.sort_by || this.config.sort_by;
      const sortDir = section.sort_dir || this.config.sort_dir;
      sectionsStates.push({
        boxes: sortBoxes(parentBoxes, calcResults.boxes, sortBy, sortDir),
        total,
        statePerPixelY: calcResults.statePerPixelY,
        spacerH: 0,
        config: section,
      });
    });

    this.sections = sectionsStates.map(sectionState => {
      // calc sizes again with the best statePerPixelY
      let totalSize = 0;
      let sizedBoxes = sectionState.boxes;
      if (sectionState.statePerPixelY !== this.statePerPixelY) {
        sizedBoxes = sizedBoxes.map(box => {
          const size = Math.max(this.config.min_box_height, Math.floor(box.state / this.statePerPixelY));
          totalSize += size;
          return {
            ...box,
            size,
          };
        });
      } else {
        totalSize = sizedBoxes.reduce((sum, b) => sum + b.size, 0);
      }
      // calc vertical margin size
      const extraSpace = this.config.height - totalSize;
      const spacerH = sizedBoxes.length > 1 ? extraSpace / (sizedBoxes.length - 1) : this.config.height;
      let offset = 0;
      // calc y positions. needed for connectors
      sizedBoxes = sizedBoxes.map(box => {
        const top = offset;
        offset += box.size + spacerH;
        return {
          ...box,
          top,
        };
      });
      return {
        ...sectionState,
        boxes: sizedBoxes,
        spacerH,
      };
    });
  }

  private _calcBoxHeights(
    boxes: Box[],
    availableHeight: number,
    totalState: number,
  ): { boxes: Box[]; statePerPixelY: number } {
    const statePerPixelY = totalState / availableHeight;
    if (statePerPixelY > this.statePerPixelY) {
      this.statePerPixelY = statePerPixelY;
    }
    let deficitHeight = 0;
    const result = boxes.map(box => {
      if (box.size === this.config.min_box_height) {
        return box;
      }
      let size = Math.floor(box.state / this.statePerPixelY);
      if (size < this.config.min_box_height) {
        deficitHeight += this.config.min_box_height - size;
        size = this.config.min_box_height;
      }
      return {
        ...box,
        size,
      };
    });
    if (deficitHeight > 0) {
      return this._calcBoxHeights(result, availableHeight - deficitHeight, totalState);
    }
    return { boxes: result, statePerPixelY: this.statePerPixelY };
  }

  private highlightPath(entityConf: EntityConfigInternal, direction: 'parents' | 'children') {
    this.highlightedEntities.push(entityConf);
    if (direction === 'children') {
      this.connectionsByParent.get(entityConf)?.forEach(c => {
        c.highlighted = true;
        this.highlightPath(c.child, 'children');
      });
    } else {
      this.connectionsByChild.get(entityConf)?.forEach(c => {
        c.highlighted = true;
        this.highlightPath(c.parent, 'parents');
      });
    }
  }

  private _handleBoxTap(box: Box): void {
    handleAction(this, this.hass, box.config, 'tap');
  }

  private _handleBoxDoubleTap(box: Box): void {
    handleAction(this, this.hass, box.config, 'double_tap');
  }

  private _handleMouseEnter(box: Box): void {
    this.highlightPath(box.config, 'children');
    this.highlightPath(box.config, 'parents');
    // trigger rerender
    this.highlightedEntities = [...this.highlightedEntities];
  }

  private _handleMouseLeave(): void {
    this.highlightedEntities = [];
    this.connections.forEach(c => {
      c.highlighted = false;
    });
  }

  private _getUnitOfMeasurement(reported_unit_of_measurement: string): string {
    // If converting to money, don't actually display the word "monetary"
    if (this.config.convert_units_to == 'monetary') {
      return '';
    }

    // If converting from kWh to gCO2, attributes.unit_of_measurement remains kWh even though the number is gCO2, so we
    // override the unit to gCO2, unless normalizeStateValue() has already converted it to kgCO2.
    if (this.config.convert_units_to && !reported_unit_of_measurement.endsWith(this.config.convert_units_to)) {
      return this.config.convert_units_to;
    }

    return reported_unit_of_measurement;
  }

  private _getEntityState(entityConf: EntityConfigInternal) {
    if (entityConf.type === 'remaining_parent_state') {
      const connections = this.connectionsByChild.get(entityConf);
      if (!connections) {
        throw new Error('Invalid entity config ' + JSON.stringify(entityConf));
      }
      const { parent } = connections[0];
      const state = connections.reduce((sum, c) => (c.ready ? sum + c.state : Infinity), 0);
      const parentEntity = this._getEntityState(parent);
      const { unit_of_measurement } = normalizeStateValue(
        this.config.unit_prefix,
        0,
        this._getUnitOfMeasurement(parentEntity.attributes.unit_of_measurement),
      );
      return { ...parentEntity, state, attributes: { ...parentEntity.attributes, unit_of_measurement } };
    }
    if (entityConf.type === 'remaining_child_state') {
      const connections = this.connectionsByParent.get(entityConf);
      if (!connections) {
        throw new Error('Invalid entity config ' + JSON.stringify(entityConf));
      }
      const { child } = connections[0];
      const state = connections.reduce((sum, c) => (c.ready ? sum + c.state : Infinity), 0);
      const childEntity = this._getEntityState(child);
      const { unit_of_measurement } = normalizeStateValue(
        this.config.unit_prefix,
        0,
        this._getUnitOfMeasurement(childEntity.attributes.unit_of_measurement),
      );
      return { ...childEntity, state, attributes: { ...childEntity.attributes, unit_of_measurement } };
    }

    let entity = this.states[getEntityId(entityConf)];
    if (!entity) {
      throw new Error('Entity not found "' + getEntityId(entityConf) + '"');
    }

    if (entityConf.attribute) {
      entity = { ...entity, state: entity.attributes[entityConf.attribute] } as HassEntity;
      if (entityConf.unit_of_measurement) {
        entity = {
          ...entity,
          attributes: { ...entity.attributes, unit_of_measurement: entityConf.unit_of_measurement },
        };
      }
    }
    return entity;
  }

  static get styles(): CSSResultGroup {
    return styles;
  }

  // https://lit.dev/docs/components/rendering/
  protected render(): TemplateResult | void {
    try {
      if (this.error) {
        throw this.error;
      }
      this.entityStates.clear();
      const containerClasses = classMap({
        container: true,
        wide: !!this.config.wide,
        'with-header': !!this.config.title,
      });

      if (!Object.keys(this.states).length) {
        return html`
          <ha-card label="Sankey Chart" .header=${this.config.title}>
            <div class=${containerClasses} style=${styleMap({ height: this.config.height + 'px' })}>
              ${localize('common.loading')}
            </div>
          </ha-card>
        `;
      }

      this._calcConnections();
      this._calcBoxes();

      this.lastUpdate = Date.now();

      return html`
        <ha-card label="Sankey Chart" .header=${this.config.title}>
          <div class=${containerClasses} style=${styleMap({ height: this.config.height + 'px' })}>
            ${this.sections.map((s, i) =>
              renderSection({
                locale: this.hass.locale,
                config: this.config,
                section: s,
                nextSection: this.sections[i + 1],
                sectionIndex: i,
                highlightedEntities: this.highlightedEntities,
                statePerPixelY: this.statePerPixelY,
                connectionsByParent: this.connectionsByParent,
                connectionsByChild: this.connectionsByChild,
                onTap: this._handleBoxTap.bind(this),
                onDoubleTap: this._handleBoxDoubleTap.bind(this),
                onMouseEnter: this._handleMouseEnter.bind(this),
                onMouseLeave: this._handleMouseLeave.bind(this),
              }),
            )}
          </div>
        </ha-card>
      `;
    } catch (err) {
      console.error(err);
      return html`${until(renderError(String(err), this.config, this.hass))}`;
    }
  }
}

export default Chart;
