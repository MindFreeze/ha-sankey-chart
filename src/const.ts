import { EntityConfig } from "./types";

export const UNIT_PREFIXES = {
  'm': 0.001,
  'k': 1000,
  'M': 1000000,
  'G': 1000000000,
  'T': 1000000000000,
};

export const MIN_LABEL_HEIGHT = 15;

export const DEFAULT_ENTITY_CONF: Omit<EntityConfig, 'entity_id'> = {
  type: 'entity',
};

export const FT3_PER_M3 = 35.31;
