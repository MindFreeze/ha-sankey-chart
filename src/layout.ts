import { MIN_HORIZONTAL_SECTION_W } from './const';
import { Config } from './types';

export function shouldBeVertical(config: Config, width: number) {
  if (config.layout === 'auto') {
    const minWidth = config.sections.reduce((acc, section) => {
      const width = section.min_width ?? MIN_HORIZONTAL_SECTION_W;
      return width + acc;
    }, 0);
    return width < minWidth;
  }
  return config.layout === 'vertical';
}
