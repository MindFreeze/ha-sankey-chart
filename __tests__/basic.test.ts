// import '../dist/ha-sankey-chart';
import { HomeAssistant } from 'custom-card-helpers';
import '../src/ha-sankey-chart';
import '../src/chart';
import SankeyChart from '../src/ha-sankey-chart';
import { SankeyChartConfig } from '../src/types';
import mockHass from './__mocks__/hass.mock';
import { LitElement } from 'lit';

const hass = mockHass({
  ent1: {
    entity_id: 'ent1',
    state: '2',
    attributes: {
      unit_of_measurement: 'W',
    },
  },
  ent2: {
    entity_id: 'ent2',
    state: '1',
    attributes: {
      unit_of_measurement: 'W',
    },
  },
  ent3: {
    entity_id: 'ent3',
    state: '1',
    attributes: {
      unit_of_measurement: 'W',
    },
  },
});

describe('SankeyChart', () => {
  const ROOT_TAG = 'sankey-chart';
  let sankeyChart: SankeyChart;

  beforeEach(() => {
    sankeyChart = window.document.createElement(ROOT_TAG) as SankeyChart;
    // @ts-ignore
    sankeyChart.hass = hass as HomeAssistant;
  });

  afterEach(() => {
    sankeyChart.remove();
  });

  it('matches a simple snapshot', async () => {
    const config: SankeyChartConfig = {
      type: '',
      sections: [
        {
          entities: [
            {
              entity_id: 'ent1',
              children: ['ent2', 'ent3'],
            },
          ],
        },
        { entities: ['ent2', 'ent3'] },
      ],
    };
    sankeyChart.setConfig(config, true);
    document.body.appendChild(sankeyChart);
    await sankeyChart.updateComplete;
    expect(sankeyChart.shadowRoot?.innerHTML.replace(/<!--.*-->/g, '')).toMatchSnapshot();

    // Wait for the sankey-chart-base component to finish updating
    const sankeyChartBase = sankeyChart.shadowRoot?.querySelector('sankey-chart-base') as LitElement;
    expect(sankeyChartBase).not.toBeNull();

    expect(sankeyChartBase.shadowRoot?.innerHTML.replace(/<!--.*-->/g, '')).toMatchSnapshot();
  });
});
