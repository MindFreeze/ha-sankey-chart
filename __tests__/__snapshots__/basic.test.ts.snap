// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SankeyChart matches a simple snapshot 1`] = `
"
      <sankey-chart-base></sankey-chart-base>

      
    "
`;

exports[`SankeyChart matches a simple snapshot 2`] = `
"
        <ha-card label="Sankey Chart">
          <div class=" container " style="height:200px;">
            
    <div class="section" style="">
      <div class="connectors">
            <svg preserveAspectRatio="none" viewBox="0 0 100 200">
        <defs>
          
            <linearGradient id="gradient0.0.0" gradientTransform="">
              <stop offset="0%" stop-color="var(--primary-color)"></stop>
              <stop offset="100%" stop-color="var(--primary-color)"></stop>
            </linearGradient>
          
            <linearGradient id="gradient0.0.1" gradientTransform="">
              <stop offset="0%" stop-color="var(--primary-color)"></stop>
              <stop offset="100%" stop-color="var(--primary-color)"></stop>
            </linearGradient>
          
        </defs>
        
              <path d="M0,2.5 C50,2.5 50,0 100,0 L100,97 C50,97 50,100 0,100 Z" fill="url(#gradient0.0.0)" fill-opacity="0.4"></path>
            
              <path d="M0,100 C50,100 50,103 100,103 L100,200 C50,200 50,197.5 0,197.5 Z" fill="url(#gradient0.0.1)" fill-opacity="0.4"></path>
            
      </svg>
          </div>
      
          <div class="spacerv" style="height:2.5px;"></div>
          
          <div class="box type-entity" style="height:195px;">
            <div style="background-color:var(--primary-color);" title="2W " class="">
              
            </div>
            <div class="label" style="line-height:15px;">
    <span>
          <span class="state">W</span>
        </span>
    
  </div>
          </div>
          
        
    </div>
  
    <div class="section" style="">
      
      
          
          
          <div class="box type-entity" style="height:97px;">
            <div style="background-color:var(--primary-color);" title="1W " class="">
              
            </div>
            <div class="label" style="line-height:15px;">
    <span>
          <span class="state">W</span>
        </span>
    
  </div>
          </div>
          
        
          <div class="spacerv" style="height:6px;"></div>
          
          <div class="box type-entity" style="height:97px;">
            <div style="background-color:var(--primary-color);" title="1W " class="">
              
            </div>
            <div class="label" style="line-height:15px;">
    <span>
          <span class="state">W</span>
        </span>
    
  </div>
          </div>
          
        
    </div>
  
          </div>
        </ha-card>
      <style>
    ha-card {
        overflow-x: auto;
    }
    .container {
        display: flex;
        position: relative;
        /* height: 210px; */
        padding: 16px;
        overflow: hidden;
    }
    .container.with-header {
        margin-top: -16px;
    }
    .container.vertical {
        flex-direction: column;
    }
    .section {
        flex: 1;
        flex-direction: column;
        position: relative;
        min-width: 0;
        max-width: 50%;
    }
    .wide .section:last-child {
        flex: initial;
    }
    .vertical .section {
        display: flex;
        flex: initial;
        flex-direction: row-reverse;
        align-items: flex-start;
        max-width: 100%;
        width: 100%;
        height: 150px;
    }
    .vertical .section:last-child {
        flex: 1;
    }
    .spacerv {
        transition: height 0.25s;
    }
    .vertical .spacerv {
        transition: width 0.25s;
    }
    .box {
        display: flex;
        align-items: center;
        /* position: relative; */
        /* min-height: 1px; */
        transition: height 0.25s;
    }
    .vertical .box {
        flex-direction: column;
        transition: width 0.25s;
    }
    /* .box::before {
        content: "";
        position: absolute;
        top: -2px;
        bottom: -2px;
        left: -2px;
        right: -2px;
        background-color: var(--primary-color);
        opacity: 0.5;
        border-radius: 3px;
    } */
    .box div:first-child {
        display: flex;
        justify-content: center;
        align-items: center;
        overflow: hidden;
        background-color: var(--primary-color);
        width: 15px;
        height: 100%;
        cursor: pointer;
    }
    .vertical .box div:first-child {
        width: 100%;
        height: 15px;
    }
    .box.type-passthrough div:first-child {
        opacity: 0.4;
    }
    .box.type-passthrough div.hl:first-child {
        opacity: 0.85;
    }
    .box .label {
        flex: 1;
        display: flex;
        align-items: center;
        padding: 0 10px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
    }
    .vertical .box .label {
        padding: 5px 0 0;
        flex-direction: column;
        white-space: normal;
        /* word-break: break-all; */
        text-align: center;
    }
    .box .label .name {
        font-style: italic;
        font-size: inherit;
    }
    .box .label .name a {
        position: sticky;
        z-index: 100;
        color: var(--primary-text-color);
    }
    .connectors {
        position: absolute;
        top: 0;
        left: 15px;
        right: 0;
        height: 100%;
        overflow: hidden;
    }
    .vertical .connectors {
        top: 15px;
        left: 0;
        bottom: 0;
        height: auto;
    }
    .connectors svg {
        position: absolute;
        left: -1px;
        width: 101%;
        height: 100%;
    }
    .vertical .connectors svg {
        top: -1px;
        left: 0;
        width: 100%;
        height: 101%;
    }
</style>"
`;
